cmd/spire-server/cli/agent/clean.go

```diff
@@ -0,0 +1,154 @@
+package agent
+
+import (
+	"flag"
+	"fmt"
+	"time"
+
+	"github.com/mitchellh/cli"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	agentv1 "github.com/spiffe/spire-api-sdk/proto/spire/api/server/agent/v1"
+	"github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+	"github.com/spiffe/spire/cmd/spire-server/util"
+	commoncli "github.com/spiffe/spire/pkg/common/cli"
+	"github.com/spiffe/spire/pkg/common/cliprinter"
+	"github.com/spiffe/spire/pkg/common/idutil"
+	"golang.org/x/net/context"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+type cleanCommand struct {
+	env           *commoncli.Env
+	expiredBefore string
+	dryRun        bool
+	printer       cliprinter.Printer
+}
+
+func NewCleanCommand() cli.Command {
+	return NewCleanCommandWithEnv(commoncli.DefaultEnv)
+}
+
+func NewCleanCommandWithEnv(env *commoncli.Env) cli.Command {
+	return util.AdaptCommand(env, &cleanCommand{env: env})
+}
+
+func (*cleanCommand) Name() string {
+	return "agent clean"
+}
+
+func (*cleanCommand) Synopsis() string {
+	return "Delete expired agents that attested using a non-TOFU security model based on a given time"
+}
+
+func (c *cleanCommand) Run(ctx context.Context, _ *commoncli.Env, serverClient util.ServerClient) (err error) {
+	expiredBefore, err := c.parseExpiredBefore()
+	if err != nil {
+		return err
+	}
+
+	agentClient := serverClient.NewAgentClient()
+	resp, err := agentClient.ListAgents(ctx, &agentv1.ListAgentsRequest{
+		Filter:     &agentv1.ListAgentsRequest_Filter{ByCanReattest: wrapperspb.Bool(true)},
+		OutputMask: &types.AgentMask{X509SvidExpiresAt: true},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to list agents: %w", err)
+	}
+
+	agents := resp.GetAgents()
+	expiredAgents := &ExpiredAgents{Agents: []*ExpiredAgent{}}
+
+	for _, agent := range agents {
```

Comment:
I think this is fine for now, though my preference would be for the prune logic to happen in SPIRE core, and expose an RPC. In the future, I think we'll want to regularly call this prune logic against agents that we know are ~safe to flush out